### DISCUSSION TITLE
Synchronous and asynchronous exceptions

### DISCUSSION THREAD

**Discussion by amichai-bd** [FPGA-MAFIA/fpga_mafia#476] [General]
Synchronous and asynchronous exceptions are fundamental concepts in computer architecture, distinguishing between exceptions caused directly by the execution of instructions and those resulting from external events or conditions. Here's a breakdown of examples for both types:

### Synchronous Exceptions
Synchronous exceptions, also known as traps, are directly caused by the execution of the current instruction sequence. They include:

1. **Instruction Address Misaligned**: Occurs when an instruction is fetched from a non-aligned address that does not conform to the architecture's alignment requirements.
2. **Load Address Misaligned**: Similar to the instruction address misalignment, but occurs during a load operation from a non-aligned memory address.
3. **Store/AMO Address Misaligned**: Occurs during a store or atomic memory operation to a non-aligned memory address.
4. **Page Fault**: Occurs when accessing a page that is not currently mapped in the memory management unit (MMU), requiring a page table walk or indicating a protection violation.
5. **Floating Point Exception**: Arises when a floating-point operation results in an undefined or unrepresentable result, such as division by zero, overflow, or underflow.
6. **System Call**: A deliberate exception triggered by software to request a service from the operating system's kernel.

### Asynchronous Exceptions (Interrupts)
Asynchronous exceptions, or interrupts, occur independently of the instruction stream currently being executed. They are typically triggered by external events. Examples include:

1. **External Interrupts**: Generated by devices outside the CPU, such as network cards, keyboards, or other peripherals, signaling that they require attention.
2. **Timer Interrupts**: Generated by a timer within the processor, useful for operating system schedulers to implement multitasking by periodically interrupting the current process.
3. **Inter-Processor Interrupts (IPIs)**: Used in multi-processor systems where one processor needs to interrupt or signal another processor, often for task scheduling or synchronization purposes.

Both synchronous and asynchronous exceptions are integral to the operation of modern computing systems, enabling efficient and controlled management of resources, error handling, and interaction between hardware and software components.

**Comment by amichai-bd** [comment]
@roman012285 - this should be on the documentation.

I really think this is a great explanation

**Reply by roman012285** [reply]
Very nice, I will add it tommorow
